Push the runtime filter from HashJoin down to SeqScan.

zhangyue-hashdata opened the pull request:
Push the runtime filter from HashJoin down to SeqScan.

+----------+  AttrFilter   +------+  ScanKey   +---------+
| HashJoin | ------------> | Hash | ---------> | SeqScan |
+----------+               +------+            +---------+

If "gp_enable_runtime_filter_pushdown" is on, three steps will be run:

Step 1. In ExecInitHashJoin(), try to find the mapper between the var in
        hashclauses and the var in SeqScan. If found we will save the mapper in
        AttrFilter and push them to Hash node;

Step 2. We will create the range/bloom filters in AttrFilter during building
        hash table, and these filters will be converted to the list of ScanKey
        and pushed down to Seqscan when the building finishes;

Step 3. ScanKeys will be used to filter slot in Seqscan.

TODO:
1. support singlenode or utility mode;
2. support null value filter;
3. support Motion, SharedScan as target node;
4. support join qual like: t1.c1 = t2.c1+5;

perf:
CPU E5-2680 v2 10 cores, memory 32GB, 3 segments
1. tpcds 10s off: 939s on: 779s 17%
2. tpcds 100s off: 5270s on: 4365s 17%

tpcds 10s details
| NO.  |    off      |    on    |     |
|------|------------|-----------|-----|
|1|2959.5|1312.9|
|2|4522.2|2995.9|
|3|11924.1|9170.6|
|4|1678.4|1653.4|
|5-1|17433.1|14723.7|
|5-2|17244.1|14499.7|
|6|5541.8|4443.8|
|7|3144.0|1726.7|
|8|3895.7|2010.0|
|9|5991.1|3270.7|
|10|29981.7|19975.5|
|11|3113.9|2293.0|
|12|2166.9|1543.3|
|13|1258.8|726.0|
|14|11745.1|7037.3|
|15|3878.6|2568.8|
|16|5420.8|3200.1|
|17|3291.2|2117.3|
|18|2103.7|979.8|
|19|4400.1|2371.1|
|20|14048.3|11135.6|
|21|7739.5|7489.6|
|22|5524.1|3083.1|
|23|5978.4|4041.7|
|24|2355.7|1993.0|
|25|5561.1|3355.0|
|26|6973.0|3180.3|
|27|3536.0|2134.0|
|28|6899.6|3141.5|
|29|6065.5|3766.1|
|30|15443.3|15137.7|
|31|9197.8|6387.3|
|32|3345.3|1815.2|
|33|2488.1|2451.7|
|34|10840.0|9291.0|
|35|52991.2|52504.5|
|36|3386.1|3381.4|
|37|1834.0|994.2|
|38|10047.6|8915.8|
|39|3035.8|1961.5|
|40|1591.5|795.5|
|41|9284.0|7623.0|
|42|10255.5|10051.7|
|43|45356.5|44808.1|
|44|1989.3|1015.8|
|45|3021.2|1222.5|
|46|19865.7|18714.8|
|47|8438.0|4972.7|
|48|4520.7|3117.3|
|49|6244.7|6150.8|
|50|5926.0|4983.4|
|51|20423.2|20287.7|
|52|84.5|85.5|
|53|9273.4|6599.0|
|54|10214.5|8094.6|
|55|38920.5|36340.8|
|56|4248.9|2761.0|
|57|1980.7|1264.9|
|58|3954.6|1806.1|
|59|2982.5|1234.5|
|60|4099.5|1852.3|
|61|3080.7|1237.5|
|62|167.6|162.6|
|63|691.0|661.1|
|64|1283.0|549.9|
|65|2033.5|1017.7|
|66|8816.5|5342.4|
|67|10124.9|6664.9|
|68-1|53775.3|53665.4|
|68-2|55517.4|54534.5|
|69-1|20955.6|18578.4|
|69-2|17315.6|15543.9|
|70|6269.5|6134.9|
|71|13798.9|12300.0|
|72|5199.1|2861.3|
|73|4017.0|2236.1|
|74|5578.1|5341.3|
|75|1056.9|687.7|
|76|16922.5|16252.2|
|77|10491.0|8369.5|
|78|4590.9|2885.0|
|79|3660.0|2081.5|
|80|967.8|516.1|
|81|6620.6|1254.1|
|82|3072.2|1214.4|
|83|5111.3|2710.2|
|84|7458.4|7442.9|
|85|2989.0|1955.6|
|86|2546.8|1522.6|
|87|45634.8|43461.0|
|88|3555.6|1858.1|
|89|5630.1|4286.9|
|90|5321.3|2992.1|
|91|4748.6|4522.8|
|92-1|6264.3|4859.7|
|92-2|6355.4|5334.0|
|93|31174.3|30890.3|
|94|3971.3|3955.2|
|95|5097.4|3004.3|
|96|1999.7|1100.9|
|97|5055.0|2974.5|
|98|7319.9|5527.3|
|99|1895.3|1838.4|
| total|  939       |  779      |   ~17%  |

tpcds 100s details

| NO.  |    off      |    on    |     |
|------|------------|-----------|-----|
|1|18410.5|8966.4|
|2|29278.7|17942.0|
|3|68332.8|52880.0|
|4|9753.3|9171.7|
|5-1|13129.4|11131.0|
|5-2|13198.6|11041.2|
|6|33375.2|27429.7|
|7|19269.8|10077.7|
|8|22145.1|12336.3|
|9|64808.8|56969.8|
|10|241285.3|161582.3|
|11|12075.9|10860.6|
|12|9563.9|8114.5|
|13|2687.4|1484.8|
|14|8932.8|5325.5|
|15|28111.9|20017.6|
|16|30523.0|19822.1|
|17|3168.7|3126.4|
|18|8063.8|5925.4|
|19|19826.5|10147.2|
|20|85671.7|80051.7|
|21|42448.4|38791.5|
|22|27746.9|15857.7|
|23|38362.9|25001.5|
|24|14143.9|13982.5|
|25|38145.1|18836.1|
|26|21448.6|12437.5|
|27|22101.6|14043.4|
|28|20168.0|12036.7|
|29|25105.1|13663.3|
|30|129563.3|129378.8|
|31|13768.9|7181.7|
|32|19039.5|11568.3|
|33|10040.4|9460.8|
|34|66348.7|57294.4|
|35|202825.5|197756.5|
|36|28703.6|28536.2|
|37|7065.0|6596.1|
|38|82447.6|73465.7|
|39|16568.5|10866.6|
|40|5632.2|2801.4|
|41|65710.0|54836.4|
|42|63198.1|61843.4|
|43|154716.5|151916.7|
|44|10148.0|5314.8|
|45|18080.4|8917.2|
|46|92069.5|84997.4|
|47|29166.7|21165.4|
|48|23630.6|20430.5|
|49|43623.1|43465.4|
|50|33574.3|24155.7|
|51|140174.1|136511.3|
|52|121.1|108.4|
|53|61777.5|53161.0|
|54|82215.5|70357.6|
|55|28835.3|26623.2|
|56|24767.9|15975.6|
|57|11274.8|7005.6|
|58|20008.8|11185.9|
|59|19055.5|8743.5|
|60|30663.6|13416.8|
|61|18951.8|8765.2|
|62|205.8|170.3|
|63|22338.1|13134.9|
|64|5539.4|3077.0|
|65|9817.0|5337.0|
|66|52164.9|26808.1|
|67|20848.3|9974.1|
|68-1|462977.0|458799.8|
|68-2|468924.8|464181.1|
|69-1|109512.7|103476.2|
|69-2|96662.2|88990.1|
|70|29996.5|28361.0|
|71|90191.5|75458.1|
|72|35578.6|20232.3|
|73|23311.9|13611.2|
|74|33299.4|30755.5|
|75|4249.5|3892.5|
|76|107381.3|102310.4|
|77|83335.9|72471.8|
|78|24223.1|15703.0|
|79|22379.5|14098.0|
|80|2753.5|1177.6|
|81|39643.7|15849.0|
|82|18515.7|8568.5|
|83|21661.9|11949.9|
|84|55856.0|55736.4|
|85|15100.4|9251.1|
|86|14383.5|9051.7|
|87|136009.8|134747.4|
|88|20819.7|12613.0|
|89|40187.3|31457.0|
|90|34545.7|12278.3|
|91|34661.6|32747.8|
|92-1|102286.6|31937.6|
|92-2|103260.1|31959.5|
|93|240605.7|235582.7|
|94|26150.7|25874.4|
|95|31199.3|20272.8|
|96|4154.5|2226.1|
|97|26195.2|17051.1|
|98|37977.6|24553.3|
|99|16624.3|16603.3|
| total|  5270       |  4365      |   ~17%  |

<!-- Thank you for your contribution to Apache Cloudberry (incubating)! -->

Fixes #ISSUE_Number

### What does this PR do?
<!-- Brief overview of the changes, including any major features or fixes -->

### Type of Change
- [ ] Bug fix (non-breaking change)
- [ ] New feature (non-breaking change)
- [ ] Breaking change (fix or feature with breaking changes)
- [ ] Documentation update

### Breaking Changes
<!-- Remove if not applicable. If yes, explain impact and migration path -->

### Test Plan
<!-- How did you test these changes? -->
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] Passed `make installcheck`
- [ ] Passed `make -C src/test installcheck-cbdb-parallel`

### Impact
<!-- Remove sections that don't apply -->
**Performance:**
<!-- Any performance implications? -->

**User-facing changes:**
<!-- Any changes visible to users? -->

**Dependencies:**
<!-- New dependencies or version changes? -->

### Checklist
- [ ] Followed [contribution guide](https://cloudberry.apache.org/contribute/code)
- [ ] Added/updated documentation
- [ ] Reviewed code for security implications
- [ ] Requested review from [cloudberry committers](https://github.com/orgs/apache/teams/cloudberry-committers)

### Additional Context
<!-- Any other information that would help reviewers? Remove if none -->

⚠️ **To skip CI:** Add `[skip ci]` to your PR title. Only use when necessary! ⚠️

---
<!-- Join our community:
- Mailing list: [dev@cloudberry.apache.org](https://lists.apache.org/list.html?dev@cloudberry.apache.org) (subscribe: dev-subscribe@cloudberry.apache.org)
- Discussions: https://github.com/apache/cloudberry/discussions -->

yjhjstz commented on src/backend/executor/nodeHash.c:
```diff
@@ -4126,3 +4151,138 @@ get_hash_mem(void)
+
+		/* append new runtime filters to target node */
+		SeqScanState *sss = castNode(SeqScanState, attr_filter->target);
+		sss->filters = list_concat(sss->filters, scankeys);
```
can we merge filter here on the same attno ?

zhangyue-hashdata replied:
1. Combining Bloom filters will result in a higher False Positive Rate (FPR) compared to using each of the individual Bloom filters separately, so it is not recommended;
2. There is the same problem to combine range filters like combining Bloom filters;
3. There is only one Bloom filter and one range filter on the same attribute in many cases;

yjhjstz replied:
```sql
create table rt2(a int, b int);
create table rt3(a int, b int);
insert into t1 select i, i from generate_series(1, 100000) i;
insert into t1 select i, i+1 from generate_series(1, 10) i;
insert into rt1 select i, i+1 from generate_series(1, 10) i;
insert into rt2 select i, i+1 from generate_series(1, 10000) i;
insert into rt3 select i, i+1 from generate_series(1, 10) i;
analyze t1;
analyze rt1;
analyze rt2;
analyze rt3;

explain analyze select * from rt1 join t1 on rt1.a = t1.b join rt3 on rt3.a = t1.b;

postgres=# explain select * from rt1 join t1 on rt1.a = t1.b join rt3 on rt3.a = t1.b;
                                   QUERY PLAN                                   
--------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=2.45..428.51 rows=17 width=24)
   ->  Hash Join  (cost=2.45..428.29 rows=6 width=24)
         Hash Cond: (t1.b = rt1.a)
         ->  Hash Join  (cost=1.23..427.00 rows=6 width=16)
               Hash Cond: (t1.b = rt3.a)
               ->  Seq Scan on t1  (cost=0.00..342.37 rows=33337 width=8)
               ->  Hash  (cost=1.10..1.10 rows=10 width=8)
                     ->  Seq Scan on rt3  (cost=0.00..1.10 rows=10 width=8)
         ->  Hash  (cost=1.10..1.10 rows=10 width=8)
               ->  Seq Scan on rt1  (cost=0.00..1.10 rows=10 width=8)
 Optimizer: Postgres query optimizer
(11 rows)

```
you can try this case, will got two range filters.

zhangyue-hashdata replied:
got it

zhangyue-hashdata replied:
It may be a better choice to let the user of the runtime filter decide whether to merge. :-)

yjhjstz commented on src/backend/executor/nodeSeqscan.c:
```diff
@@ -365,3 +398,60 @@ ExecSeqScanInitializeWorker(SeqScanState *node,
+			continue;
+
+		val = slot_getattr(slot, sk->sk_attno, &isnull);
+		if (isnull)
```
```c++
CREATE TABLE distinct_1(a int);
CREATE TABLE distinct_2(a int);
INSERT INTO distinct_1 VALUES(1),(2),(NULL);
INSERT INTO distinct_2 VALUES(1),(NULL);
SELECT * FROM distinct_1, distinct_2 WHERE distinct_1.a IS NOT DISTINCT FROM distinct_2.a;
```
test got wrong result.

zhangyue-hashdata replied:
I will fix it.

yjhjstz commented on src/backend/executor/nodeSeqscan.c:
```diff
@@ -87,8 +101,17 @@ SeqNext(SeqScanState *node)
+			return slot;
+
+		if (node->filter_in_seqscan && node->filters &&
+			!PassByBloomFilter(node, slot))
```
tpcds 1TB, bloom filter will lose efficacy or create failed due to large rows ?

zhangyue-hashdata replied:
Yes, when creating the Bloom filter, the system evaluates the estimated number of rows that this hash join will process and the amount of available memory during the execution plan generation. It determines whether using a Bloom filter for filtering data would be effective based on this evaluation. If it is assessed that the Bloom filter would not sufficiently enhance performance, then the Bloom filter will not be created.

fanfuxiaoran replied:
> It determines whether using a Bloom filter for filtering data would be effective based on this evaluation

That makes sense, but where is related code, I just didn't see them in this pr. 
Does it compares the number of rows between the output of hashtable and data in the probe table? If the rows of the hashtable are far less than that of the probe table , then use the runtime filter?

yjhjstz commented on src/test/regress/expected/gp_runtime_filter.out:
```diff
@@ -250,6 +250,60 @@ SELECT COUNT(*) FROM dim_rf
+         Extra Text: (seg2)   Hash chain length 8.0 avg, 8 max, using 4 of 524288 buckets.
+         ->  Seq Scan on t1 (actual rows=1 loops=1)
+         ->  Hash (actual rows=32 loops=1)
+               Buckets: 524288  Batches: 1  Memory Usage: 4098kB
```
how to debug and get pushdown scankey here ?

gfphoenix78 replied:
The test case should cover that the hash join node or result node is the child of the parent hash join.

gfphoenix78 replied:
> how to debug and get pushdown scankey here ?

Add debug message to dump the stats about how many tuples are filter out?

zhangyue-hashdata replied:
Add debug message in https://github.com/apache/cloudberry/pull/724/commits/274d8aa1ff0b33a4d24e3a7765b493904c61dd60

fanfuxiaoran commented on src/backend/executor/nodeHashjoin.c:
```diff
@@ -2157,3 +2174,225 @@ ExecHashJoinInitializeWorker(HashJoinState *state,
+	{
+		match = false;
+
+		if (!IsA(lfirst(lc), Var))
```
Could it support other expression, whose one arg is the column attr, and the other is a const?

zhangyue-hashdata replied:
I think that expressions like `t1.c1 = 5` should be pushed down by the optimizer to operators such as SeqScan for early processing. Therefore, this feature does not handle expressions of the form `t1.c1 = 5`.

fanfuxiaoran replied:
Sorry, I didn't make it clear.  I don't mean the predication on the var.  like the below sql
```
 EXPLAIN (ANALYZE, COSTS OFF, SUMMARY OFF, TIMING OFF)
SELECT t1.c3 FROM t1, t2 WHERE t1.c2 = (t2.c2 + 10);
                                        QUERY PLAN
-------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3) (actual rows=0 loops=1)
   ->  Hash Join (actual rows=0 loops=1)
         Hash Cond: (t1.c2 = (t2.c2 + 10))
         Extra Text: (seg2)   Hash chain length 8.0 avg, 8 max, using 4 of 524288 buckets.
         ->  Seq Scan on t1 (actual rows=128 loops=1)
         ->  Hash (actual rows=32 loops=1)
               Buckets: 524288  Batches: 1  Memory Usage: 4098kB
               ->  Seq Scan on t2 (actual rows=32 loops=1)
 Optimizer: Postgres query optimizer
(9 rows)
```
As `t2.c2 + 10` is not a `Var` but a `T_OpExpr` , the runtime filter cannot handle it.
Could we just iterate the expression tree and check if it only contains var and const ?

zhangyue-hashdata replied:
I got it. This scenario should be supported at a later date. However, since this is a type of scenario that might require a significant amount of code to implement, I think it would be best to support it at an appropriate time in the future.

fanfuxiaoran commented:
Looks interesting. And I have some questions to discuss.
- Beside the seqscan, can the runtime filter apply to other types of scan? such as the index scan.
- Looks only when the `hashjoin` node and `seqscan` node run in the same process can use the runtime filter. Which means the tables should have same distributed policy on the join columns or one of the table is replicated.

yjhjstz commented on src/backend/executor/nodeHashjoin.c:
```diff
@@ -2157,3 +2174,225 @@ ExecHashJoinInitializeWorker(HashJoinState *state,
+		 *   result (hash filter)
+		 *     seqscan on t1, t1 is replicated table
+		 */
+		if (!IsA(child, HashJoinState) && !IsA(child, ResultState))
```
```c++
Hash Join  (cost=0.00..4019.55 rows=37 width=9) (actual time=3203.012..9927.435 rows=1399 loops=1)
                                                   Hash Cond: (web_sales_1_prt_2.ws_item_sk = item.i_item_sk)
                                                   Join Filter: (web_sales_1_prt_2.ws_ext_discount_amt > ((1.3 * avg(web_sales_1_prt_2_1.ws_ext_discount_amt))))
                                                   Rows Removed by Join Filter: 4763
                                                   Extra Text: (seg2)   Hash chain length 1.0 avg, 1 max, using 198 of 2097152 buckets.
                                                   ->  Append  (cost=0.00..676.44 rows=2399189 width=13) (actual time=16.899..5572.473 rows=3090021 loops=1)
                                                         ->  Seq Scan on web_sales_1_prt_2  (cost=0.00..676.44 rows=2399189 width=13) (actual time=16.895..1138.267 rows=662
149 loops=1)
                                                         ->  Seq Scan on web_sales_1_prt_3  (cost=0.00..676.44 rows=2399189 width=13) (actual time=8.947..1102.409 rows=6621
36 loops=1)
                                                         ->  Seq Scan on web_sales_1_prt_4  (cost=0.00..676.44 rows=2399189 width=13) (actual time=8.822..1100.839 rows=6621
48 loops=1)
                                                         ->  Seq Scan on web_sales_1_prt_5  (cost=0.00..676.44 rows=2399189 width=13) (actual time=11.391..1083.785 rows=662
179 loops=1)
                                                         ->  Seq Scan on web_sales_1_prt_6  (cost=0.00..676.44 rows=2399189 width=13) (actual time=13.030..649.141 rows=4414
09 loops=1)
                                                         ->  Seq Scan on web_sales_1_prt_7  (cost=0.00..676.44 rows=2399189 width=13) (never executed)
                                                         ->  Seq Scan on web_sales_1_prt_others  (cost=0.00..676.44 rows=2399189 width=13) (actual time=1.213..3.203 rows=17
88 loops=1)
                                                   ->  Hash  (cost=2432.09..2432.09 rows=109 width=12) (actual time=3177.768..3177.770 rows=198 loops=1)
                                                         Buckets: 2097152  Batches: 1  Memory Usage: 16392kB
                                                         ->  Broadcast Motion 3:3  (slice3; segments: 3)  (cost=
```
need to consider partitioned table .

zhangyue-hashdata replied:
I will try.

zhangyue-hashdata replied:
support partitioned table in https://github.com/apache/cloudberry/pull/724/commits/c70109269c71e616734a1083159eaf44d6362e48

zhangyue-hashdata commented:
> Looks interesting. And I have some questions to discuss.
> 
> * Beside the seqscan, can the runtime filter apply to other types of scan? such as the index scan.
> * Looks only when the `hashjoin` node and `seqscan` node run in the same process can use the runtime filter. Which means the tables should have same distributed policy on the join columns or one of the table is replicated.

gfphoenix78 commented on src/backend/executor/nodeHash.c:
```diff
@@ -4126,3 +4151,138 @@ get_hash_mem(void)
+
+		attr_filter->blm_filter     = bloom_create_aggresive(node->ps.plan->plan_rows,
+		                                             work_mem, random());
+		attr_filter->min    = LLONG_MAX;
+		attr_filter->max    = LLONG_MIN;
```
LLONG_MAX, LLONG_MIN are platform-spec value, i.e. the bound value for `unsigned long long`, which may not be exactly the same width as Datum. For safety, static assert could be considered.

zhangyue-hashdata replied:
I see `StaticAssertDecl(SIZEOF_DATUM == 8, "sizeof datum is not 8");` in postgres.h, so it's better to use INT64_MAX/INT64_MIN here.

yjhjstz replied:
use  LONG_MAX, LONG_MIN instead ?

zhangyue-hashdata replied:
fix with https://github.com/apache/cloudberry/pull/724/commits/99eabb224f89dcdfd7117ef37a5986ab8001abdf

gfphoenix78 commented on src/backend/executor/nodeHashjoin.c:
```diff
@@ -2157,3 +2174,225 @@ ExecHashJoinInitializeWorker(HashJoinState *state,
+	AttrFilter	*attr_filter;
+	ListCell	*lc;
+
+	/*
+	 * Only applicatable for inner, right and semi join,
+	 */
```
Could you give a little more explain about why these join types are supported and others are not?

zhangyue-hashdata replied:
Add more message to explain why just inner, right and semi join are allowed with runtime filter.
fix it in https://github.com/apache/cloudberry/commit/98dac6dfc7d5e44e111aa16bdf5948d07ee2eb00

gfphoenix78 commented on src/backend/executor/nodeHashjoin.c:
```diff
@@ -2157,3 +2174,225 @@ ExecHashJoinInitializeWorker(HashJoinState *state,
+	if (lattno == NULL || rattno == NULL)
+		return false;
+
+	if (!IsA(expr, OpExpr) && !IsA(expr, FuncExpr))
+		return false;
```
These 2 lines duplicate with the following if-elseif-else code, could be deleted.

zhangyue-hashdata replied:
fix it

zhangyue-hashdata replied:
fix with https://github.com/apache/cloudberry/pull/724/commits/99eabb224f89dcdfd7117ef37a5986ab8001abdf

gfphoenix78 commented on src/backend/executor/nodeHashjoin.c:
```diff
@@ -2157,3 +2174,225 @@ ExecHashJoinInitializeWorker(HashJoinState *state,
+		match = false;
+
+		if (!IsA(lfirst(lc), Var))
+			break;
+
+		var = lfirst(lc);
+		if (var->varno == INNER_VAR)
+			*rattno = var->varattno;
+		else if (var->varno == OUTER_VAR)
+			*lattno = var->varattno;
+		else
+			break;
+
+		match = true;
+	}
+
+	return match;
```
The match flag gets the code hard(several modifications) to read. The break statement could be replaced by `return false;`. If the foreach loop ends, all conditions match, so returns true.

gfphoenix78 commented on src/backend/executor/nodeSeqscan.c:
```diff
@@ -87,8 +101,17 @@ SeqNext(SeqScanState *node)
-	if (table_scan_getnextslot(scandesc, direction, slot))
+	while (table_scan_getnextslot(scandesc, direction, slot))
+	{
+		if (TupIsNull(slot))
+			return slot;
```
Will it be true?

zhangyue-hashdata replied:
It seems that slot is never NULL here, so `Assert(!TupIsNull(slot));` is better or remove them.

zhangyue-hashdata replied:
fix with https://github.com/apache/cloudberry/pull/724/commits/99eabb224f89dcdfd7117ef37a5986ab8001abdf

gfphoenix78 commented on src/backend/executor/nodeSeqscan.c:
```diff
@@ -365,3 +398,60 @@ ExecSeqScanInitializeWorker(SeqScanState *node,
+	for (int i = 0; i < *num; ++i)
+		memcpy(&sk[i], list_nth(keys, i), sizeof(ScanKeyData));
+
+	/*
+	 * SK_EMPYT means the end of the array of the ScanKey
+	 */
+	sk[*num].sk_flags = SK_EMPYT;
```
How to check the boundary of the `ScanKey` array in rescan? In normal rescan, the number of `ScanKey`s is the same as begin_scan. If the number of `ScanKey`s is larger in rescan than that in begin_scan, the boundary value might be invalid and dangerous to access.

zhangyue-hashdata replied:
Currently, there are no tables in CBDB that support runtime filter pushdown to AM. So runtime filter pushdown to AM will not be supported temporarily, and filtering operations can be performed only in SeqScan.

yjhjstz commented on src/backend/executor/nodeSeqscan.c:
```diff
@@ -87,8 +101,17 @@ SeqNext(SeqScanState *node)
+			return slot;
+
+		if (node->filter_in_seqscan && node->filters &&
+			!PassByBloomFilter(node, slot))
```
what's difference with `gp_enable_runtime_filter`'s bloom filter ?

zhangyue-hashdata replied:
**Similarities:**
In this feature, the creation and testing of the Bloom filter use the same implementation as that of the gp_enable_runtime_filter feature.

**Differences:**

- The gp_enable_runtime_filter feature decides during the optimizer phase whether to use a Bloom filter to filter data. It inserts the functionality of filtering data with a Bloom filter as an nodeRuntimeFilter node into the left subtree of a Hash Join.

- In contrast, this new feature does not insert an node. Instead, it directly pushes down the Bloom filter and other filters to the SeqScan or even the Access Method (AM), applying the filters closer to where the data is generated. This approach enhances filtering efficiency compared to gp_enable_runtime_filter.

By pushing the filters closer to the data source, this new method achieves more efficient data filtering before the data reaches higher-level operators, potentially reducing the amount of data that needs to be processed by subsequent operations in the query execution plan.

zhangyue-hashdata commented on src/backend/executor/nodeHashjoin.c:
```diff
@@ -162,6 +164,16 @@ static void ReleaseHashTable(HashJoinState *node);
+static void CreateRuntimeFilter(HashJoinState* hjstate);
+static bool IsEqualOp(Expr *expr);
+static bool CheckEqualArgs(Expr *expr, AttrNumber *lattno, AttrNumber *rattno);
+static PlanState *FindTargetAttr(HashJoinState *hjstate,
```
good idea, i will fix it.

avamingli commented:
There are codes changed  in MultiExecParallelHash, please add some parallel tests with runtime filter.

zhangyue-hashdata commented:
> There are codes changed in MultiExecParallelHash, please add some parallel tests with runtime filter.

got it.

zhangyue-hashdata commented:
> Looks interesting. And I have some questions to discuss.
> 
> * Beside the seqscan, can the runtime filter apply to other types of scan? such as the index scan.
> * Looks only when the `hashjoin` node and `seqscan` node run in the same process can use the runtime filter. Which means the tables should have same distributed policy on the join columns or one of the table is replicated.

* Beside the seqscan, can the runtime filter apply to other types of scan? such as the index scan.
> Theoretically, it is feasible to apply runtime filters to operators such as Index Scan. However, because Index Scan already reduces data volume by leveraging an optimized storage structure, the performance gains from applying runtime filters to Index Scan would likely be minimal. Thus, I think that applying runtime filters to Index Scan would not yield significant performance benefits.

> In subsequent work, when we discover that other scan operators can achieve notable performance improvements from pushdown runtime filters, we will support these operators. Our focus will be on operators where runtime filters can substantially decrease the amount of data processed early in the query execution, leading to more pronounced performance enhancements.

* Looks only when the `hashjoin` node and `seqscan` node run in the same process can use the runtime filter. Which means the tables should have same distributed policy on the join columns or one of the table is replicated.
> Yes, the current pushdown runtime filter only supports in-process pushdown, which means that the Hash Join and SeqScan need to be within the same process. The design and implementation of cross-process pushdown runtime filters are much more complex.

> This limitation arises because coordinating and sharing data structures like Bloom filters or other runtime filters across different processes involves additional challenges such as inter-process communication (IPC), synchronization, and ensuring consistency and efficiency of the filters across process boundaries. Addressing these issues requires a more sophisticated design that can handle the complexities of distributed computing environments.

avamingli commented:
Hi, with gp_enable_runtime_filter_pushdown = on, execute SQL below will get a crash:

```sql
gpadmin=# show gp_enable_runtime_filter_pushdown;
 gp_enable_runtime_filter_pushdown
-----------------------------------
 on
(1 row)
```
```sql
CREATE TABLE test_tablesample (dist int, id int, name text) WITH (fillfactor=10) DISTRIBUTED BY (dist);
-- use fillfactor so we don't have to load too much data to get multiple pages

-- Changed the column length in order to match the expected results based on relation's blocksz
INSERT INTO test_tablesample SELECT 0, i, repeat(i::text, 875) FROM generate_series(0, 9) s(i) ORDER BY i;
INSERT INTO test_tablesample SELECT 3, i, repeat(i::text, 875) FROM generate_series(10, 19) s(i) ORDER BY i;
INSERT INTO test_tablesample SELECT 5, i, repeat(i::text, 875) FROM generate_series(20, 29) s(i) ORDER BY i;
EXPLAIN (COSTS OFF)
  SELECT id FROM test_tablesample TABLESAMPLE SYSTEM (50) REPEATABLE (2);
FATAL:  Unexpected internal error (assert.c:48)
DETAIL:  FailedAssertion("IsA(planstate, SeqScanState)", File: "explain.c", Line: 4154)
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Succeeded.
psql (14.4, server 14.4)
```

<img width="1779" alt="image" src="https://github.com/user-attachments/assets/a8882208-1800-4df3-9f2d-7c659e24aaa3">

zhangyue-hashdata commented:
> ```sql
> ```sql
> gpadmin=# show gp_enable_runtime_filter_pushdown;
>  gp_enable_runtime_filter_pushdown
> -----------------------------------
>  on
> (1 row)
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> ```sql
> CREATE TABLE test_tablesample (dist int, id int, name text) WITH (fillfactor=10) DISTRIBUTED BY (dist);
> -- use fillfactor so we don't have to load too much data to get multiple pages
> 
> -- Changed the column length in order to match the expected results based on relation's blocksz
> INSERT INTO test_tablesample SELECT 0, i, repeat(i::text, 875) FROM generate_series(0, 9) s(i) ORDER BY i;
> INSERT INTO test_tablesample SELECT 3, i, repeat(i::text, 875) FROM generate_series(10, 19) s(i) ORDER BY i;
> INSERT INTO test_tablesample SELECT 5, i, repeat(i::text, 875) FROM generate_series(20, 29) s(i) ORDER BY i;
> EXPLAIN (COSTS OFF)
>   SELECT id FROM test_tablesample TABLESAMPLE SYSTEM (50) REPEATABLE (2);
> FATAL:  Unexpected internal error (assert.c:48)
> DETAIL:  FailedAssertion("IsA(planstate, SeqScanState)", File: "explain.c", Line: 4154)
> server closed the connection unexpectedly
>         This probably means the server terminated abnormally
>         before or while processing the request.
> The connection to the server was lost. Attempting reset: Succeeded.
> psql (14.4, server 14.4)
> ```
> ```

Thanks, I'll reproduce the issue and fix it.

fanfuxiaoran commented:
Thanks for your detailed explanation. 

> > Looks interesting. And I have some questions to discuss.
> > 
> > * Beside the seqscan, can the runtime filter apply to other types of scan? such as the index scan.
> > * Looks only when the `hashjoin` node and `seqscan` node run in the same process can use the runtime filter. Which means the tables should have same distributed policy on the join columns or one of the table is replicated.
> 
> * Beside the seqscan, can the runtime filter apply to other types of scan? such as the index scan.
> 
> > Theoretically, it is feasible to apply runtime filters to operators such as Index Scan. However, because Index Scan already reduces data volume by leveraging an optimized storage structure, the performance gains from applying runtime filters to Index Scan would likely be minimal. Thus, I think that applying runtime filters to Index Scan would not yield significant performance benefits.
> 
Make sense. When doing hashjoin, index scan or index only scan are often not used on probe node.

> > In subsequent work, when we discover that other scan operators can achieve notable performance improvements from pushdown runtime filters, we will support these operators. Our focus will be on operators where runtime filters can substantially decrease the amount of data processed early in the query execution, leading to more pronounced performance enhancements.
> 
> * Looks only when the `hashjoin` node and `seqscan` node run in the same process can use the runtime filter. Which means the tables should have same distributed policy on the join columns or one of the table is replicated.
> 
> > Yes, the current pushdown runtime filter only supports in-process pushdown, which means that the Hash Join and SeqScan need to be within the same process. The design and implementation of cross-process pushdown runtime filters are much more complex.
> 
> > This limitation arises because coordinating and sharing data structures like Bloom filters or other runtime filters across different processes involves additional challenges such as inter-process communication (IPC), synchronization, and ensuring consistency and efficiency of the filters across process boundaries. Addressing these issues requires a more sophisticated design that can handle the complexities of distributed computing environments.

Exactly, and if  there is any lock  used to solve the problem may even lead bad performance.

fanfuxiaoran commented:
```
 explain analyze
SELECT count(t1.c3) FROM t1, t3 WHERE t1.c1 = t3.c1 ;
                                                              QUERY PLAN

-----------------------------------------------------------------------------------------
----------------------------------------------
 Finalize Aggregate  (cost=1700.07..1700.08 rows=1 width=8) (actual time=32119.566..32119
.571 rows=1 loops=1)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=1700.02..1700.07 rows=3 width=8) (
actual time=30.967..32119.550 rows=3 loops=1)
         ->  Partial Aggregate  (cost=1700.02..1700.03 rows=1 width=8) (actual time=32119
.131..32119.135 rows=1 loops=1)
               ->  Hash Join  (cost=771.01..1616.68 rows=33334 width=4) (actual time=14.0
59..32116.962 rows=33462 loops=1)
                     Hash Cond: (t3.c1 = t1.c1)
                     Extra Text: (seg0)   Hash chain length 1.0 avg, 3 max, using 32439 o
f 524288 buckets.
                     ->  Seq Scan on t3  (cost=0.00..387.34 rows=33334 width=4) (actual t
ime=0.028..32089.490 rows=33462 loops=1)
                     ->  Hash  (cost=354.34..354.34 rows=33334 width=8) (actual time=13.2
57..13.259 rows=33462 loops=1)
                           Buckets: 524288  Batches: 1  Memory Usage: 5404kB
                           ->  Seq Scan on t1  (cost=0.00..354.34 rows=33334 width=8) (ac
tual time=0.180..4.877 rows=33462 loops=1)
 Planning Time: 0.227 ms
```
runtime_filter has been pushed down to t3 table seqscan, but 'explain analyze' doesn't print them out.

```
\d t1
                 Table "public.t1"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 c1     | integer |           |          |
 c2     | integer |           |          |
 c3     | integer |           |          |
 c4     | integer |           |          |
 c5     | integer |           |          |
Checksum: t
Indexes:
    "t1_c2" btree (c2)
Distributed by: (c1)
```
```
 \d t3
                 Table "public.t3"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 c1     | integer |           |          |
 c2     | integer |           |          |
 c3     | integer |           |          |
 c4     | integer |           |          |
 c5     | integer |           |          |
Distributed by: (c1)
```

zhangyue-hashdata commented:
> ```
>  explain analyze
> SELECT count(t1.c3) FROM t1, t3 WHERE t1.c1 = t3.c1 ;
>                                                               QUERY PLAN
> 
> -----------------------------------------------------------------------------------------
> ----------------------------------------------
>  Finalize Aggregate  (cost=1700.07..1700.08 rows=1 width=8) (actual time=32119.566..32119
> .571 rows=1 loops=1)
>    ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=1700.02..1700.07 rows=3 width=8) (
> actual time=30.967..32119.550 rows=3 loops=1)
>          ->  Partial Aggregate  (cost=1700.02..1700.03 rows=1 width=8) (actual time=32119
> .131..32119.135 rows=1 loops=1)
>                ->  Hash Join  (cost=771.01..1616.68 rows=33334 width=4) (actual time=14.0
> 59..32116.962 rows=33462 loops=1)
>                      Hash Cond: (t3.c1 = t1.c1)
>                      Extra Text: (seg0)   Hash chain length 1.0 avg, 3 max, using 32439 o
> f 524288 buckets.
>                      ->  Seq Scan on t3  (cost=0.00..387.34 rows=33334 width=4) (actual t
> ime=0.028..32089.490 rows=33462 loops=1)
>                      ->  Hash  (cost=354.34..354.34 rows=33334 width=8) (actual time=13.2
> 57..13.259 rows=33462 loops=1)
>                            Buckets: 524288  Batches: 1  Memory Usage: 5404kB
>                            ->  Seq Scan on t1  (cost=0.00..354.34 rows=33334 width=8) (ac
> tual time=0.180..4.877 rows=33462 loops=1)
>  Planning Time: 0.227 ms
> ```
> 
> runtime_filter has been pushed down to t3 table seqscan, but 'explain analyze' doesn't print them out.
> 
> ```
> \d t1
>                  Table "public.t1"
>  Column |  Type   | Collation | Nullable | Default
> --------+---------+-----------+----------+---------
>  c1     | integer |           |          |
>  c2     | integer |           |          |
>  c3     | integer |           |          |
>  c4     | integer |           |          |
>  c5     | integer |           |          |
> Checksum: t
> Indexes:
>     "t1_c2" btree (c2)
> Distributed by: (c1)
> ```
> 
> ```
>  \d t3
>                  Table "public.t3"
>  Column |  Type   | Collation | Nullable | Default
> --------+---------+-----------+----------+---------
>  c1     | integer |           |          |
>  c2     | integer |           |          |
>  c3     | integer |           |          |
>  c4     | integer |           |          |
>  c5     | integer |           |          |
> Distributed by: (c1)
> ```

Thanks for your test case. Based on these, I rewrote code to ensure that debug info are always displayed even when the number of filtered rows is zero. And add the test case into gp_runtime_filter.sql too.
fix in https://github.com/apache/cloudberry/commit/98dac6dfc7d5e44e111aa16bdf5948d07ee2eb00

zhangyue-hashdata commented:
> Hi, with gp_enable_runtime_filter_pushdown = on, execute SQL below will get a crash:
> 
> ```sql
> gpadmin=# show gp_enable_runtime_filter_pushdown;
>  gp_enable_runtime_filter_pushdown
> -----------------------------------
>  on
> (1 row)
> ```
> 
> ```sql
> CREATE TABLE test_tablesample (dist int, id int, name text) WITH (fillfactor=10) DISTRIBUTED BY (dist);
> -- use fillfactor so we don't have to load too much data to get multiple pages
> 
> -- Changed the column length in order to match the expected results based on relation's blocksz
> INSERT INTO test_tablesample SELECT 0, i, repeat(i::text, 875) FROM generate_series(0, 9) s(i) ORDER BY i;
> INSERT INTO test_tablesample SELECT 3, i, repeat(i::text, 875) FROM generate_series(10, 19) s(i) ORDER BY i;
> INSERT INTO test_tablesample SELECT 5, i, repeat(i::text, 875) FROM generate_series(20, 29) s(i) ORDER BY i;
> EXPLAIN (COSTS OFF)
>   SELECT id FROM test_tablesample TABLESAMPLE SYSTEM (50) REPEATABLE (2);
> FATAL:  Unexpected internal error (assert.c:48)
> DETAIL:  FailedAssertion("IsA(planstate, SeqScanState)", File: "explain.c", Line: 4154)
> server closed the connection unexpectedly
>         This probably means the server terminated abnormally
>         before or while processing the request.
> The connection to the server was lost. Attempting reset: Succeeded.
> psql (14.4, server 14.4)
> ```
> 
> <img alt="image" width="1779" src="https://private-user-images.githubusercontent.com/17311022/392676960-a8882208-1800-4df3-9f2d-7c659e24aaa3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM0MTAzMTUsIm5iZiI6MTczMzQxMDAxNSwicGF0aCI6Ii8xNzMxMTAyMi8zOTI2NzY5NjAtYTg4ODIyMDgtMTgwMC00ZGYzLTlmMmQtN2M2NTllMjRhYWEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA1VDE0NDY1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhYzExMjYyNmM3MTllZGExNWJhMmZhNjRiZjYwODg4MmY3YTVlMjcyZjYxMjg2YzVkZmM1MmY4OTNlZWQxNTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.inCY4gvv-RdOzeQPxitJ7obOz2FFpSMeYCgWzrERhS0">

hanks for your test case. I fix it in https://github.com/apache/cloudberry/commit/98dac6dfc7d5e44e111aa16bdf5948d07ee2eb00 And add the test case into gp_runtime_filter.sql too.
fix in https://github.com/apache/cloudberry/commit/98dac6dfc7d5e44e111aa16bdf5948d07ee2eb00

Smyatkin-Maxim commented:
Hi @zhangyue-hashdata 
I see that previous runtime filter implementation relies on some cost model at try_runtime_filter(). Do I understand it correctly, that this PR does not do any cost evaluation?
Also for TPC-H/TPC-DS can you provide results for each query separately?

Asking mostly out of curiosity, I see here are quite a few reviewers here already :)

zhangyue-hashdata commented:
> Hi @zhangyue-hashdata I see that previous runtime filter implementation relies on some cost model at try_runtime_filter(). Do I understand it correctly, that this PR does not do any cost evaluation? Also for TPC-H/TPC-DS can you provide results for each query separately?
> 
> Asking mostly out of curiosity, I see here are quite a few reviewers here already :)

Basically, you're correct. Because our goal is to filter out as much data as possible right at the point of data generation. However, this will lead to very complex evaluations, so we only made a simple estimation based on rows and work memory when creating the Bloom filter.
Furthermore, I have placed the detailed test results for TPC-DS 10s in PR description.

zhangyue-hashdata commented:
> There are codes changed in MultiExecParallelHash, please add some parallel tests with runtime filter.

fix it in https://github.com/apache/cloudberry/pull/724/commits/7ab040ae178e9bb616bd75e0488aa6a2293d4183

yjhjstz commented on src/backend/executor/nodeSeqscan.c:
```diff
@@ -87,8 +101,14 @@ SeqNext(SeqScanState *node)
-	if (table_scan_getnextslot(scandesc, direction, slot))
+	while (table_scan_getnextslot(scandesc, direction, slot))
+	{
+		if (node->filter_in_seqscan && node->filters &&
```
```c++
if (!node->filter_in_seqscan ||  !node->filters)
{
     if (table_scan_getnextslot(scandesc, direction, slot))
         return slot;
}
else 
{
     while (table_scan_getnextslot(scandesc, direction, slot))
     {
            .....
     }
}
```

this make origin path more efficient and readable ?

zhangyue-hashdata replied:
good idea! I fix in https://github.com/apache/cloudberry/pull/724/commits/bcf93e66f72a35db0ab5694a92c59c768423b1ab

yjhjstz commented:
from tpcds 10s details table, there are some bad cases.

zhangyue-hashdata commented:
> from tpcds 10s details table, there are some bad cases.

**21,24,30,42,49,54,68-1,99**

I retested these SQL statements that exhibited performance regression, and the latest test results show no noticeable performance difference when toggling **gp_enable_runtime_filter_pushdown**. So, I speculate that the performance regression in these SQL statements might be associated with testing method. Previously I tested by running the entire suite of 99 TPC-DS queries with **gp_enable_runtime_filter_pushdown** enabled, and then again with it disabled.

Therefore, a more appropriate method would be to execute the same SQL statement multiple times with **gp_enable_runtime_filter_pushdown** both enabled and disabled, respectively, and then take the average of those runs for comparison. I will follow this testing method for retesting and observe if there's any performance regression.

zhangyue-hashdata commented:
> > from tpcds 10s details table, there are some bad cases.
> 
> **21,24,30,42,49,54,68-1,99**
> 
> I retested these SQL statements that exhibited performance regression, and the latest test results show no noticeable performance difference when toggling **gp_enable_runtime_filter_pushdown**. So, I speculate that the performance regression in these SQL statements might be associated with testing method. Previously I tested by running the entire suite of 99 TPC-DS queries with **gp_enable_runtime_filter_pushdown** enabled, and then again with it disabled.
> 
> Therefore, a more appropriate method would be to execute the same SQL statement multiple times with **gp_enable_runtime_filter_pushdown** both enabled and disabled, respectively, and then take the average of those runs for comparison. I will follow this testing method for retesting and observe if there's any performance regression.

I have retested the performance of tpcds 10s using the previously mentioned testing method. Please see the description part for the latest results.

yjhjstz commented:
> I have retested the performance of tpcds 10s using the previously mentioned testing method. Please see the description part for the latest results

cool, what about tpcds 100 sf ?

zhangyue-hashdata commented:
> > I have retested the performance of tpcds 10s using the previously mentioned testing method. Please see the description part for the latest results
> 
> cool, what about tpcds 100 sf ?

Please see the description part for the results of tpcds 100s.

yjhjstz commented on src/backend/executor/nodeSeqscan.c:
```diff
@@ -87,8 +100,22 @@ SeqNext(SeqScanState *node)
+	{
+		while (table_scan_getnextslot(scandesc, direction, slot))
+		{
+			if (!PassByBloomFilter(node, slot))
```
Is bloom filter works on tpc-ds 1000G ?

zhangyue-hashdata replied:
Based on the test results of TPC-DS 10G and 100G, I estimate that there is a high possibility of observing a similar performance improvement margin in the TPC-DS 1000G performance test as well.

yjhjstz replied:
but bloom_create_aggresive maybe return NULL, buffer limit 1M ~2MB

zhangyue-hashdata replied:
> but bloom_create_aggresive maybe return NULL, buffer limit 1M ~2MB

There are two factors that will ensure the Bloom filter is created successfully with a high probability. The first is that Bloom filters are typically built on small tables. Even for tables as large as 1TB or 10TB, the data volume of TPC-DS small tables is generally not very large. The second factor is the number of segments; the average amount of data from small tables distributed across each segment will also not be excessive. Considering these two factors together, for a data volume of 10TB, it is highly likely that the creation of the Bloom filter will not fail.

yjhjstz commented on src/backend/executor/nodeSeqscan.c:
```diff
@@ -244,12 +281,27 @@ void
 	scan = node->ss.ss_currentScanDesc;
 
+	/*
+	 * Clear all the pushdown scan keys.
```
why clear all ?

zhangyue-hashdata replied:
I revert the fix with https://github.com/apache/cloudberry/pull/724/commits/05cfbdbc25f50f6ec164fb7229ee8140a27de84f

Smyatkin-Maxim commented on src/backend/executor/nodeHashjoin.c:
```diff
@@ -2156,3 +2176,293 @@ ExecHashJoinInitializeWorker(HashJoinState *state,
+		if (lattno == -1 || targets == NULL)
+			continue;
+
+		foreach(lc, targets)
```
Is having the same listcell variable here intended?
I guess it should be working, but i`m not sure how it gonna behave if at some point we add a break statement into the inner loop. And I'm not sure if this code will continue working in future PG releases :)

zhangyue-hashdata replied:
> Is having the same listcell variable here intended? I guess it should be working, but i`m not sure how it gonna behave if at some point we add a break statement into the inner loop. And I'm not sure if this code will continue working in future PG releases :)

At the commit **Support partition table**, I changed the return type of `FindTargetNodes()` from `PlanState *` to `List *` for partition table.

Smyatkin-Maxim commented:
@zhangyue-hashdata have you tried benchmarks on the recent builds? The last time I ran TPC-H against cloudberry, 12/22 queries used legacy optimizer. I believe now, after cherry-picks from gpdb master, more queries should be using ORCA optimizer. I wonder if there is still the same benefit from runtime bloom filters.

Btw, can you guys share TPC-DS and TPC-H toolkit you're using to benchmark cloudberry?

zhangyue-hashdata commented:
> @zhangyue-hashdata have you tried benchmarks on the recent builds? The last time I ran TPC-H against cloudberry, 12/22 queries used legacy optimizer. I guess for TPC-DS it's even worse. But now, after cherry-picks from gpdb master, more queries should be using ORCA optimizer. I wonder if there is still the same benefit from runtime bloom filters.
> 
> Btw, can you guys share TPC-DS and TPC-H toolkit you're using to benchmark cloudberry?

not yet recently, here is toolkit for TPC-DS, hope helpful to you :-)
[toolkit.tar.gz](https://github.com/user-attachments/files/19459216/toolkit.tar.gz)

gfphoenix78 commented on src/backend/executor/nodeHash.c:
```diff
@@ -511,6 +523,9 @@ ExecEndHash(HashState *node)
 	outerPlan = outerPlanState(node);
 	ExecEndNode(outerPlan);
+
+	if (gp_enable_runtime_filter_pushdown && node->filters)
```
Is the if condition necessary here? We should always free the filter if the filter is not null

zhangyue-hashdata replied:
fix in [3828a4d](https://github.com/apache/cloudberry/pull/724/commits/3828a4d743d77c0c39501b8419640065291c343a)

gfphoenix78 commented on src/backend/executor/nodeHashjoin.c:
```diff
@@ -999,6 +1014,11 @@ ExecInitHashJoin(HashJoin *node, EState *estate, int eflags)
 		ExecInitRuntimeFilterFinish(rfstate, hstate->ps.plan->plan_rows);
 	}
 
+	if (Gp_role == GP_ROLE_EXECUTE
+		&& gp_enable_runtime_filter_pushdown
+		&& !estate->useMppParallelMode)
+		CreateRuntimeFilter(hjstate);
```
Will we support runtime filter in singlenode or utility mode?

gfphoenix78 replied:
If yes, we should also add tests in src/test/singlenode_regress/

zhangyue-hashdata replied:
This support will be added to the TODO list.

gfphoenix78 commented on src/backend/executor/nodeHash.c:
```diff
@@ -4125,3 +4143,142 @@ get_hash_mem(void)
+}
+
+static void
+BuildRuntimeFilter(HashState *node, TupleTableSlot *slot)
```
The function name is not very good to me. `BuildXXX` looks do something only one pass, but it actually runs for every tuples.

zhangyue-hashdata replied:
`AddTupleValuesIntoRF` should be better, I fix in [3828a4d](https://github.com/apache/cloudberry/pull/724/commits/3828a4d743d77c0c39501b8419640065291c343a)

gfphoenix78 reviewed (approved):
LGTM

yjhjstz reviewed (approved):
LGTM

yjhjstz commented:
@zhangyue-hashdata squash commits into one commit .
